packages/react/focus-scope/src/FocusScope.tsx

```diff
@@ -26,44 +35,265 @@ type FocusScopeProps = {
+
+/* -------------------------------------------------------------------------------------------------
+ * Utils
+ * -----------------------------------------------------------------------------------------------*/
+
+/**
+ * Attempts focusing the first element in a list of candidates.
+ * Stops when focus has actually moved.
+ */
+function focusFirst(candidates: HTMLElement[], { select = false } = {}) {
+  const PREVIOUSLY_FOCUSED_ELEMENT = document.activeElement;
+  for (const candidate of candidates) {
+    focus(candidate, { select });
+    if (document.activeElement !== PREVIOUSLY_FOCUSED_ELEMENT) return;
+  }
+}
+
+/**
+ * Returns the first and last tabbable elements inside a container.
+ * NOTE: This takes into account if elements are hidden or not.
+ */
+function getTabbableEdges(container: HTMLElement) {
+  const candidates = getTabbableCandidates(container);
+  const first = getFirstRealTabbable(candidates, container);
+  const last = getFirstRealTabbable(candidates.reverse(), container);
+  return [first, last] as const;
+}
+
+/**
+ * Returns a list of potential tabbable candidates.
+ * NOTE: They are considered candidates at this point, because it includes elements that are hidden.
+ */
+function getTabbableCandidates(container: HTMLElement) {
+  const nodes: HTMLElement[] = [];
+  const walker = createTabbableWalker(container);
+  while (walker.nextNode()) nodes.push(walker.currentNode as HTMLElement);
+  // we do not take into account the order of nodes with positive `tabIndex` as it is
+  // considered an accessibility issue.
+  return nodes;
+}
+
+/**
+ * Creates a tree that contains only tabbable elements.
+ * This is only a close aproximation. For example it doesn't take into account cases like when
+ * elements are not visible. This cannot be worked out easily by just reading a property, but rather
+ * necessitate runtime knowledge (computed styles, etc). We deal with these cases separately.
+ *
+ * See: https://developer.mozilla.org/en-US/docs/Web/API/TreeWalker
+ * Credit: https://github.com/discord/focus-layers/blob/master/src/util/wrapFocus.tsx#L1
+ */
+function createTabbableWalker(container: HTMLElement) {
+  return document.createTreeWalker(container, NodeFilter.SHOW_ELEMENT, {
+    acceptNode: (node: any) => {
+      const isHiddenInput = node.tagName === 'INPUT' && node.type === 'hidden';
+      if (node.disabled || node.hidden || isHiddenInput) return NodeFilter.FILTER_SKIP;
+      // `.tabIndex` is not the same as the `tabindex` attribute. It works on the
+      // runtime's understanding of tabbability, so this automatically accounts
+      // for any kind of element that could be tabbed to.
+      return node.tabIndex >= 0 ? NodeFilter.FILTER_ACCEPT : NodeFilter.FILTER_SKIP;
+    },
+  });
+}
+
+/**
+ * Returns the first "real" tabbable in a list of candidates.
+ * NOTE: This takes into account if elements are hidden or not.
+ */
+function getFirstRealTabbable(candidates: HTMLElement[], container: HTMLElement) {
```

Comment:
thoughts on renaming to `findVisibleTabbable` ? The word "real" didn't help much cos I don't immediately understand what a "fake" tabbable is haha. also, `find` always returns first in `Array.find` so that made sense too but not hugely fussed about that part.

The rename will prevent me from having to check implementation to understand.

Reply:
I like the name. Yeah I find the naming really tricky to get right because of the fact that we have tabbable that may not in fact be tabbable… annoying.
I'm tempted to call this one even just `findFirstTabbable`. Thoughts?